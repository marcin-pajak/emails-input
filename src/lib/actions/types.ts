import { Elements, Handlers } from '../components/types';
import { Options } from '../options/types';
import { State, StateAction } from '../state/types';

/**
 * Email actions
 */
export interface EmailAddActions {
  add: (value: string | string[]) => boolean;
}

export interface EmailRemoveActions {
  removeByIndex: (index: number) => void;
  removeByLocalId: (localId: string) => void;
  removeByName: (name: string) => void;
}

export type EmailActions = EmailAddActions & EmailRemoveActions;

export interface EmailActionsProperties {
  elements: Elements;
  handleAction: (action: StateAction) => string | undefined;
  options: Options;
  state: State;
}

/**
 * Library actions
 */
export interface LibraryActions {
  destroy: () => void;
  init: () => void;
}

export interface LibraryActionProperties {
  elements: Elements;
  handleAction: (action: StateAction) => string | undefined;
  handlers: Handlers;
  state: State;
}

/**
 * Actions
 */
export type Actions = EmailActions & LibraryActions;
