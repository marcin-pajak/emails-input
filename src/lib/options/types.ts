import { ChangeType } from '../constants';

export interface Options {
  initialValue: string[];
  onChange: (changed: string, type: ChangeType, newValue: string[]) => void;
}
