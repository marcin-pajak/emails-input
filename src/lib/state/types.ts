import { ActionType, LibraryState } from './constants';

/**
 * State Actions
 */
export interface AddEmailAction {
  payload: string;
  type: typeof ActionType.Add;
}

export interface RemoveEmailAction {
  payload: string;
  type: typeof ActionType.Remove;
}

export interface UpdateStateAction {
  payload: LibraryState;
  type: typeof ActionType.UpdateState;
}

export type StateAction =
  | AddEmailAction
  | RemoveEmailAction
  | UpdateStateAction;

/**
 * State
 */
export interface Entry {
  email: string;
  isValid: boolean;
}

export type LocalId = string;

export type EmailMapper = Record<string, LocalId>;

export type EntryMap = Record<string, Entry>;

export interface State {
  emails: string[];
  entries: Entry[];
  entryMap: EntryMap;
  mapper: EmailMapper;
  state: LibraryState;
}

/**
 * State Manager
 */
export interface StateManager {
  handleAction: (action: StateAction, state: State) => string | undefined;
  makeState: (initialValue: string[]) => State;
}
