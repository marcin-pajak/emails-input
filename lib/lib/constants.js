"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLASSES = exports.ChangeType = void 0;
var ChangeType;
(function (ChangeType) {
    ChangeType["Added"] = "Added";
    ChangeType["Removed"] = "Removed";
})(ChangeType = exports.ChangeType || (exports.ChangeType = {}));
exports.CLASSES = {
    CONTAINER: 'EmailsInputWrapper',
    EMAIL: 'EmailsInput-email',
    INPUT: 'EmailsInput-input',
    WRAPPER: 'EmailsInput',
};
//# sourceMappingURL=constants.js.map