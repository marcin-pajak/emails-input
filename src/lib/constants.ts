export enum ChangeType {
  Added = 'Added',
  Removed = 'Removed',
}

export const CLASSES = {
  CONTAINER: 'EmailsInputWrapper',
  EMAIL: 'EmailsInput-email',
  INPUT: 'EmailsInput-input',
  WRAPPER: 'EmailsInput',
};
