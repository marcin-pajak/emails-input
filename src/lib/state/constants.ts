export enum ActionType {
  Add = 'ADD',
  Remove = 'REMOVE',
  Initialize = 'INITIALIZE',
  Destroy = 'DESTROY',
  UpdateState = 'UPDATE_STATE',
}

export enum LibraryState {
  Initializing = 'Initializing',
  Initialized = 'Initialized',
  Destroyed = 'Destroyed',
}
