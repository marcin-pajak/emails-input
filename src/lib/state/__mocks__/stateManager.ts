import { LibraryState } from '../constants';
import { State, StateManager } from '../types';

export const stateManager = (): StateManager => ({
  handleAction: jest.fn(),
  makeState: (): State => ({
    emails: ['new@email.com'],
    entries: [
      {
        email: 'new@email.com',
        isValid: true,
      },
    ],
    entryMap: {
      '1': {
        email: 'new@email.com',
        isValid: true,
      },
    },
    mapper: {
      'new@email.com': '1',
    },
    state: LibraryState.Initialized,
  }),
});
