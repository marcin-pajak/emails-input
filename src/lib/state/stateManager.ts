import { isValidEmail } from '../../utils/isValidEmail';
import { objectValues } from '../../utils/object';
import {
  handleAddAction,
  handleRemoveAction,
  handleUpdateAction,
} from './actions';
import { ActionType, LibraryState } from './constants';
import {
  EmailMapper,
  Entry,
  EntryMap,
  State,
  StateAction,
  StateManager,
} from './types';

/**
 * Library's State Manipulation
 */
export const stateManager = (): StateManager => {
  const id = {
    current: 1,
  };

  /**
   * Return new state
   * @param initialValue
   */
  const makeState = (initialValue: string[]): State => {
    const mapper: EmailMapper = {};

    const entryMap = initialValue.reduce(
      (map: EntryMap, email: string): EntryMap => {
        const newId = String(id.current);
        map[newId] = {
          email,
          isValid: isValidEmail(email),
        };
        mapper[email] = newId;
        id.current++;

        return map;
      },
      {},
    );

    return {
      emails: initialValue,
      entries: objectValues<Entry>(entryMap),
      entryMap,
      mapper,
      state: LibraryState.Initializing,
    };
  };

  /**
   * Process action
   * @param action
   * @param state
   */
  const handleAction = (
    action: StateAction,
    state: State,
  ): string | undefined => {
    switch (action.type) {
      case ActionType.Add: {
        const newId = String(id.current);
        handleAddAction(action, state, newId);
        id.current++;

        return newId;
      }

      case ActionType.Remove: {
        handleRemoveAction(action, state);

        return undefined;
      }

      case ActionType.UpdateState: {
        handleUpdateAction(action, state);

        return undefined;
      }

      default: {
        return undefined;
      }
    }
  };

  return {
    handleAction,
    makeState,
  };
};
