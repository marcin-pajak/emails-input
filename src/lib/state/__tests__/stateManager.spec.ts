import {
  addItemAction,
  removeItemAction,
  updateStateAction,
} from '../actionCreators';
import { LibraryState } from '../constants';
import { stateManager } from '../stateManager';
import { StateAction } from '../types';

describe('stateManager makeState should', (): void => {
  it('return initial state', (): void => {
    expect.hasAssertions();
    const manager = stateManager();

    expect(manager.makeState([])).toMatchSnapshot();
  });

  it('return initial state with custom value', (): void => {
    expect.hasAssertions();
    const manager = stateManager();

    expect(manager.makeState(['good@email.com'])).toMatchSnapshot();
  });
});

describe('stateManager handleAction should', (): void => {
  it('handle wrong action', (): void => {
    expect.hasAssertions();
    const manager = stateManager();
    const state = manager.makeState([]);

    const result = manager.handleAction(
      ({ type: 'wrong' } as unknown) as StateAction,
      state,
    );

    expect(result).toBeUndefined();
    expect(state).toMatchSnapshot();
  });

  it('handle adding email', (): void => {
    expect.hasAssertions();
    const manager = stateManager();
    const state = manager.makeState([]);

    const result = manager.handleAction(addItemAction('new@email.com'), state);

    expect(result).toBe('1');
    expect(state).toMatchSnapshot();
  });

  it('handle removing email', (): void => {
    expect.hasAssertions();
    const manager = stateManager();
    const state = manager.makeState([]);

    const id = manager.handleAction(addItemAction('new@email.com'), state);
    expect(state.emails).toHaveLength(1);

    const result = manager.handleAction(removeItemAction(id as string), state);

    expect(state.emails).toHaveLength(0);
    expect(result).toBeUndefined();
    expect(state).toMatchSnapshot();
  });

  it('handle updating library state', (): void => {
    expect.hasAssertions();
    const manager = stateManager();
    const state = manager.makeState([]);

    const result = manager.handleAction(
      updateStateAction(LibraryState.Destroyed),
      state,
    );

    expect(result).toBeUndefined();
    expect(state).toMatchSnapshot();
  });
});
