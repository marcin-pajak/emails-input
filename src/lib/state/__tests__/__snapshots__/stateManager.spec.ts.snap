// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stateManager handleAction should handle adding email 1`] = `
Object {
  "emails": Array [
    "new@email.com",
  ],
  "entries": Array [
    Object {
      "email": "new@email.com",
      "isValid": true,
    },
  ],
  "entryMap": Object {
    "1": Object {
      "email": "new@email.com",
      "isValid": true,
    },
  },
  "mapper": Object {
    "new@email.com": "1",
  },
  "state": "Initializing",
}
`;

exports[`stateManager handleAction should handle removing email 1`] = `
Object {
  "emails": Array [],
  "entries": Array [],
  "entryMap": Object {},
  "mapper": Object {},
  "state": "Initializing",
}
`;

exports[`stateManager handleAction should handle updating library state 1`] = `
Object {
  "emails": Array [],
  "entries": Array [],
  "entryMap": Object {},
  "mapper": Object {},
  "state": "Destroyed",
}
`;

exports[`stateManager handleAction should handle wrong action 1`] = `
Object {
  "emails": Array [],
  "entries": Array [],
  "entryMap": Object {},
  "mapper": Object {},
  "state": "Initializing",
}
`;

exports[`stateManager makeState should return initial state 1`] = `
Object {
  "emails": Array [],
  "entries": Array [],
  "entryMap": Object {},
  "mapper": Object {},
  "state": "Initializing",
}
`;

exports[`stateManager makeState should return initial state with custom value 1`] = `
Object {
  "emails": Array [
    "good@email.com",
  ],
  "entries": Array [
    Object {
      "email": "good@email.com",
      "isValid": true,
    },
  ],
  "entryMap": Object {
    "1": Object {
      "email": "good@email.com",
      "isValid": true,
    },
  },
  "mapper": Object {
    "good@email.com": "1",
  },
  "state": "Initializing",
}
`;
